app/src/main/java/com/sdp13epfl2021/projmag/database/UserDataFirebase.kt

```diff
@@ -129,4 +137,43 @@ class UserDataFirebase(
                 .addOnFailureListener(onFailure)
         }
     }
+
+    override fun applyUnapply(
+        unapply: Boolean,
+        projectId: ProjectId,
+        onSuccess: () -> Unit,
+        onFailure: (Exception) -> Unit
+    ) {
+        getUserDoc()?.let { doc ->
+            getListOfAppliedToProjects(
+                { ls ->
+                    val temp = ls.toMutableSet()
+                    if (unapply)
+                        temp.remove(projectId)
+                    else
+                        temp.add(projectId)
+                    val newList: List<ProjectId> = temp.toList()
+                    doc.set(
+                        hashMapOf(
+                            "applied to" to newList
+                        ), SetOptions.merge()
+                    ).addOnSuccessListener { onSuccess() }.addOnFailureListener(onFailure)
+                },
+                onFailure
+            )
```

Comment:
This could be simplified to an atomic update (and avoid nested database calls), for example by this
```suggestion
            val fieldValue = if (unapply) {
                FieldValue.arrayRemove(projectId)
            } else {
                FieldValue.arrayUnion(projectId)
            }
            doc.update(APPLIED_TO_FIELD, fieldValue)
                .addOnSuccessListener { onSuccess() }
                .addOnFailureListener(onFailure)
```